### PR TITLE
Rule Manager: support resource level rule replacement in isolation module

### PR DESCRIPTION
**Pull request by liqiangz**
### Describe what this PR does / why we need it
Rule Manager: support resource level rule replacement in isolation module

### Special notes for reviews

### CODE REVIEW

File: core/isolation/rule_manager.go [hunk 1]
```diff
@@ -36,54 +36,131 @@ var (
...
-	isEqual := reflect.DeepEqual(currentRules, rules)
+
+	resRulesMap := make(map[string][]*Rule, 16)
+	for _, rule := range rules {
+		resRules, exist := resRulesMap[rule.Resource]
+		if !exist {
+			resRules = make([]*Rule, 0, 1)
+		}
+		resRulesMap[rule.Resource] = append(resRules, rule)
+	}
```
Comment (luckyxiaoqiang) [inline core/isolation/rule_manager.go:47]: We can do this without lock.

File: core/isolation/rule_manager.go [hunk 2]
```diff
@@ -36,54 +36,131 @@ var (
...
 	rwMux.Lock()
-	ruleMap = m
-	currentRules = rules
+	ruleMap = validResRulesMap
 	rwMux.Unlock()
+	currentRules = rawResRulesMap
 
 	logging.Debug("[Isolation LoadRules] Time statistic(ns) for updating isolation rule", "timeCost", util.CurrentTimeNano()-start)
-	logRuleUpdate(m)
+	logRuleUpdate(validResRulesMap)
```
Comment (luckyxiaoqiang) [inline core/isolation/rule_manager.go:89]: [Isolation LoadRules] ==> [Isolation onRuleUpdate]

Comment (liqiangz) [reply core/isolation/rule_manager.go:88]: done

File: core/isolation/rule_manager.go (hunk 1 above)
Comment (liqiangz) [reply core/isolation/rule_manager.go:44]: done

**Review by luckyxiaoqiang — APPROVED** [review]
LGTM

File: core/isolation/rule_manager.go [hunk 3]
```diff
@@ -27,63 +27,139 @@ import (
...
+		logging.Info("[Isolation] Load resource level rules is the same with current resource level rules, so ignore load operation.")
+		return false, nil
+	}
+
+	err := onResourceRuleUpdate(res, rules)
+	return true, err
+}
+
+func onResourceRuleUpdate(res string, rawResRules []*Rule) (err error) {
+	validResRules := make([]*Rule, 0, len(rawResRules))
```
Comment (sanxun0325) [inline core/isolation/rule_manager.go:123]: Use defer func() recover error might be better

File: core/isolation/rule_manager.go [hunk 4]
```diff
@@ -27,63 +27,139 @@ import (
...
+	for _, rule := range rawResRules {
+		if err := IsValidRule(rule); err != nil {
+			logging.Warn("[Isolation onResourceRuleUpdate] Ignoring invalid flow rule", "rule", rule, "reason", err.Error())
```
Comment (sanxun0325) [inline core/isolation/rule_manager.go:126]: should be `Ignoring invalid isolation rule`

**Review by louyuting — APPROVED** [review]

**Pull request merged**